src/QATools/Example/Pages/LoginPage.php

```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * This file is part of the QA-Tools library.
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ *
+ * @copyright Michael Iwersen <mi.iwersen@gmail.com>
+ * @link      https://github.com/qa-tools/behat-example
+ */
+
+namespace QATools\Example\Pages;
+
+
+use QATools\QATools\HtmlElements\Element\Form;
+use QATools\QATools\HtmlElements\TypifiedPage;
+
+/**
+ * Class LoginPage.
+ *
+ * @page-url('/')
+ */
+class LoginPage extends TypifiedPage
+{
+
+	/**
+	 * The login form.
+	 *
+	 * @var Form
+	 * @find-by('id' => 'login-form')
+	 */
+	protected $loginForm;
+
+	/**
+	 * Fill the login form with given data.
+	 *
+	 * @param string $username The username.
+	 * @param string $password The password.
+	 *
+	 * @return static
+	 */
+	public function fillForm($username, $password)
+	{
+		$this->loginForm->fill(array(
+			'username' => $username,
+			'password' => $password,
+		));
+
+		return $this;
+	}
+
+	/**
+	 * Does the login.
+	 *
+	 * @return Page
+	 */
+	public function loginSuccess()
```

Comment:
> You don't share code?

Our company is large and I have no access to repositories of other departments/teams. It takes time to get access in such a large company, its even more problematic cause our team is further outside than the others.


> I'm not sure what should in step name: the stuff that we check (that error message is present on a page) or fact, that login is ok or fail (and leave detection of fail/check state as internal code in step implementation).

Good question at all. I think we can spent ages discussing the perfect wording and its implementation but I guess `Given the user logins with 'username' and 'password'` is too complex in implementation, cause it could either return a page after the successful login or return the same page and the code has to decide if its either success or fail. From reading the second step its clear that it was a failure but the implementation of the first doesn't know that.

But maybe it could sound like
```gherkin
Given the user logins with 'username' and 'password'
And the login is failing
Then the error message should say 'invalid username or password'
```
The `and` will do the actual submit and return the same page, cause we stay on the same page and the `then` will check for the error message. With the constraints we talked about the `and` checks implicit if this is correct on constructing the login page again. But I really don't know if `given` is allowed to fail after all.

For me scenario should explain what is going to happen and also what to check. I know that test cases for testers are also written in Gherkin and the single steps executed by humans. Thats why there shouldn't be any to complex implementation behind and just simple `yes` or `no` decissions.

Maybe we should just implement both. 
* few complex steps
* many atomic steps
The user can see what is possible and has two different ways?